vernac/attributes.ml

```diff
@@ -105,16 +121,77 @@ let single_key_parser ~name ~key v prev args =
   assert_once ~name prev;
   v
 
-let bool_attribute ~name ~on ~off : bool option attribute =
-  attribute_of_list [(on, single_key_parser ~name ~key:on true);
-               (off, single_key_parser ~name ~key:off false)]
+(** [key_value_attribute ~name ~default ~keys] parses a attribute [name=key]
+  with possible [key] values in [keys], [default] is for compatibility for users
+  doing [qualif(key)] which is parsed as [qualif(key=default)] *)
+let key_value_attribute ~name ~default ~(keys : (string * 'a) list) : 'a option attribute =
+  let parser = function
+    | Some v ->
+      CErrors.user_err Pp.(str "key '" ++ str name ++ str "' has been already set.")
+    | None ->
+      begin function
+        | VernacFlagLeaf (FlagIdent b) ->
+          begin match CList.assoc_f String.equal b keys with
+            | exception Not_found ->
+              CErrors.user_err
+                Pp.(str "Invalid value '" ++ str b ++ str "' for key " ++ str name ++ fnl () ++
+                    str "use one of " ++ (pr_sequence str (List.map fst keys)))
+            | value -> value
+          end
+        | VernacFlagEmpty ->
+          default
+        | err ->
+          CErrors.user_err Pp.(str "Invalid syntax " ++ pr_vernac_flag (name, err) ++ str ", try " ++ str name ++ str "=key instead.")
+      end
+  in
+  attribute_of_list [name, parser]
+
+let bool_attribute ~name : bool option attribute =
+  let keys = ["on", true; "off", false] in
+  key_value_attribute ~name ~default:true ~keys
+
+let legacy_bool_attribute ~name ~on ~off : bool option attribute =
+  attribute_of_list
+    [(on, single_key_parser ~name ~key:on true);
+     (off, single_key_parser ~name ~key:off false)]
+
+(* important note: we use on as the default for the new bool_attribute ! *)
+let deprecated_bool_attribute_warning =
+  CWarnings.create
+    ~name:"deprecated-attribute-syntax"
+    ~category:"parsing"
+    ~default:CWarnings.Enabled
```

Comment:
AFAIK this is the default so this line is not really useful.

Reply:
Personally it doesn't bother me, but I can remove if you want.

Reply:
I don't care.